### ISSUE TITLE
many out of fuel escorts causes high CPU load

### ISSUE THREAD

**Issue by hasteech** [endless-sky/endless-sky#1684]
When large numbers of escorts (hundreds) run out of fuel in the same system as you, the AI seems to grind a lot of CPU.  The grinding stops if you give them orders.  Perhaps there are too many help requests being calculated?

**Comment by Amazinite** [comment]
A lot of your problems seem to be "I have too many ships and it's making things go slow. Help."
There's only so much the engine can do. 

If you have 100+ ships and you're experiencing slow down, that's on you for having so many ships. Either trade some smaller ones out for bigger ones or just get rid of most of them, because you should really only need a handful of properly fitted ships to do everything.

It really just seems like the problem where somebody had hundreds of freighters and was able to gain so much money that their overflowed credits when into the negative. If you're pushing the game to its limits then expect problems. Roll back a bit.

**Comment by Amazinite** [comment]
#1676, 77, 78, 82, and this one all seem like "I can't hold all these ships" problems.

**Comment by Wrzlprnft** [comment]
Performance issues are those nevertheless too

**Comment by hasteech** [comment]
Yes, it's a side effect of having a lot of ships, and yes, I can and usually do play the game with fewer ships.  These are still performance issues with fairly straightforward solutions.  Since it sounds like some people have trouble running the game with even fewer ships (dozens, presumably on slower hardware), I think these would be useful for them as well.  It's not a high priority bug or anything, but I think it's worth addressing at some point.

I can actually manage most of these with careful preparation and planning (in this case, order your ships to hold), but it's a bit silly that the hardest part of running a large fleet is working around game engine inadequacies.

**Comment by endless-sky** [comment]
Fleets of hundreds of ships are really beyond the scope of what this game is supposed to be, so slow-down that only occur when you've got that many ships are very low priority to address.

**Comment by KiLEdEnNis** [comment]
try getting 10^99 Bactrians & wonder why the game is crashing :P

**Comment by pcordes** [comment]
@hasteech: many of ES's loops are O(N^2) or O(N^3) in the number of ships.  

e.g. Every frame (60 times per second),  for each ship: filter the list of all ships to make a list of enemy ships.

Then for each weapon on this ship, see if it's ready to fire, then loop over the list of enemy ships and see if this weapon *should* fire at it.  (This is `AI::AutoFire()` in AI.cpp).

There are early-out checks based on range before doing a full collision-detection between the projected flight-path of the weapon and the outline of the target ship (a polygon ranging from 16 to 250 line segments).

Even in the best case, every ship loops over every ship (O(N^2)), checking that they're in the same system, and doing a lookup in a std::map to see if its government is hostile to the other ship's gov.  And maybe checking the range.

I played around some with re-ordering the checks to put the cheapest and most-effective ones first (e.g. to put government-relations checks after in-system and range checks, since those are cheap and can filter a lot), but I haven't profiled this and am not sure.

Building this enemy list for every ship also requires slow atomic operations, because `std::shared_ptr<>` is always thread-safe, even though I think ES doesn't need that.  It has separate threads for graphics and engine, but I don't think two threads can ever be modifying ref counts in parallel.  Anyway, it's possible to work with references to the objects in the `const std::list<shared_ptr<Ship>> &ships` function arg, because those objects can't have their refcounts decremented while AutoFire is running.

i.e. it's safe to avoid copying `shared_ptr<Ship>` by value for temporary lists of ships that are deleted when the function returns.  I did this, but like I said, haven't profiled it.  My code is still messy as I figure out the cleanest way to get what we want from C++, so it's not ready for a pull-request.  But here it is:

https://github.com/pcordes/endless-sky/blob/SIMD-mask/source/AI.cpp#L1587

I turned `enemies` into a vector of `std::reference_wrapper` instead of a vector pointers:  
`vector<reference_wrapper<const shared_ptr<Ship>>> enemies;`.

A branch with mostly just that change is https://github.com/endless-sky/endless-sky/compare/master...pcordes:shared-ptr-parent-holds-refs, but the one I linked also has lots of other stuff I've been working on (like much better SSE2 vectorization, and inlining of lots of functions that are so small that the final binary is smaller when they're inlined than when they have to be called.

(Although there are so many tiny accessor functions in ES that need to be inlined for the asm output to be efficient, it might make more sense to have the build system detect and use link-time-optimization to inline across `.cpp` files, which solves the problem.  Calling a function requires saving all floating-pointer variables from registers to memory, on x86 Linux and OS X, because the calling convention has no call-preserved XMM registers.  This makes function calls more expensive for the calling function than just the overhead of setting up args and the CALL instruction itself).

I also reduced usage of sqrt.  

I turned the Angle lookup-table into a static array class member, so the lookups can inline to  inline to very good asm code.  (And yes, I do know how to look at compiler output and decide if it's good or not: See for example [this SO answer I wrote](http://stackoverflow.com/questions/40354978/why-is-this-c-code-faster-than-my-hand-written-assembly-for-testing-the-collat))
